src/Nethermind/Nethermind.JsonRpc.TraceStore/DbPersistingBlockTracer.cs

```diff
@@ -0,0 +1,85 @@
+﻿//  Copyright (c) 2021 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+//
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+//
+
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Db;
+using Nethermind.Evm.Tracing;
+using Nethermind.Int256;
+using Nethermind.Logging;
+
+namespace Nethermind.JsonRpc.TraceStore;
+
+/// <summary>
+/// Tracer that can store traces of decorated tracer in database
+/// </summary>
+/// <typeparam name="TTrace">Trace type</typeparam>
+/// <typeparam name="TTracer">Transaction tracer type</typeparam>
+public class DbPersistingBlockTracer<TTrace, TTracer> : IBlockTracer where TTracer : class, ITxTracer
+{
+    private readonly IDb _db;
+    private readonly Func<IReadOnlyCollection<TTrace>, byte[]> _serialization;
+    private readonly IBlockTracer _blockTracer;
+    private readonly BlockTracerBase<TTrace, TTracer> _tracerWithResults;
```

Comment:
NP: Any reason why not reuse `_tracer` and remove one var?

Reply:
Interface mismatch, some methods are only allowed through interface